pages/instructorAssessmentManualGrading/instructorAssessmentManualGrading.sql

```diff
@@ -7,21 +7,92 @@ SELECT
     admin_assessment_question_number(aq.id) as number,
     ag.number AS alternative_group_number,
     (count(*) OVER (PARTITION BY ag.number)) AS alternative_group_size,
-    (SELECT COUNT(DISTINCT iq.id)
+    (SELECT COUNT(DISTINCT s.id)
      FROM
          instance_questions AS iq
          JOIN variants AS v ON (v.instance_question_id = iq.id)
          JOIN submissions AS s ON (s.variant_id = v.id)
      WHERE
          iq.assessment_question_id = aq.id) AS num_submissions,
-    (SELECT COUNT(DISTINCT iq.id)
+    (SELECT COUNT(DISTINCT s.id)
+     FROM 
+         submissions AS S
+         JOIN variants AS V ON (s.variant_id = v.id)
+         JOIN instance_questions AS IQ ON (v.instance_question_id = iq.id)
+     WHERE (s.auth_user_id, s.date) 
+        IN (
+             SELECT s.auth_user_id, MAX(s.date)
+             FROM
+                 submissions AS s
+                 JOIN variants AS v ON (s.variant_id = v.id)
+                 JOIN instance_questions AS iq ON (v.instance_question_id = iq.id)
+             WHERE
+                 iq.assessment_question_id = aq.id
+             GROUP BY s.auth_user_id
+         )
+    ) AS num_submissions_eligible,
+    (SELECT COUNT(DISTINCT IQ.id)
+     FROM 
+         submissions AS S
+         JOIN variants AS V ON (s.variant_id = v.id)
+         JOIN instance_questions AS IQ ON (v.instance_question_id = iq.id)
+     WHERE (s.auth_user_id, s.date) 
+        IN (
+             SELECT s.auth_user_id, MAX(s.date)
+             FROM
+                 submissions AS s
+                 JOIN variants AS v ON (s.variant_id = v.id)
+                 JOIN instance_questions AS iq ON (v.instance_question_id = iq.id)
+             WHERE
+                 iq.assessment_question_id = aq.id
+             GROUP BY s.auth_user_id
+         )
+         AND manual_grading_user IS NOT NULL
+         AND graded_at IS NOT NULL) AS num_graded,
+    (SELECT COUNT(DISTINCT IQ.id)
+     FROM 
+         submissions AS S
+         JOIN variants AS V ON (s.variant_id = v.id)
+         JOIN instance_questions AS IQ ON (v.instance_question_id = iq.id)
+     WHERE (s.auth_user_id, s.date) 
+        IN (
+             SELECT s.auth_user_id, MAX(s.date)
+             FROM
+                 submissions AS s
+                 JOIN variants AS v ON (s.variant_id = v.id)
+                 JOIN instance_questions AS iq ON (v.instance_question_id = iq.id)
+             WHERE
+                 iq.assessment_question_id = aq.id
+             GROUP BY s.auth_user_id
+         )
+         AND manual_grading_user IS NOT NULL
+         AND graded_at IS NULL) AS num_locked_submissions,
+    (SELECT COUNT(DISTINCT IQ.id)
+     FROM 
+         submissions AS S
+         JOIN variants AS V ON (s.variant_id = v.id)
+         JOIN instance_questions AS IQ ON (v.instance_question_id = iq.id)
+     WHERE (s.auth_user_id, s.date) 
```

Comment:
I'm not loving this method of finding `max(date)` and then re-finding the submission with the maximum date. Elsewhere we sort by `s.date DESC, s.id DESC` and then LIMIT the results: https://github.com/PrairieLearn/PrairieLearn/blob/5483eb8c5698e3b3e5b7e888c1d9b4e51434e270/sprocs/variants_select_submission_for_grading.sql#L18

Can we do the same here with a `DISTINCT ON (iq.id)` to get the list of instance questions that we are interested in?

Reply:
I think the only problem here is that we are searching for ungraded submissions, which we _then_  sort and limit by. 

Hence, if I am a student who makes multiple submissions, and then someone grades my last submission, the proposed query logic above will begin to return the second to last submission that is ungraded, even though my instance question has already been graded.

I really don't like the subquery either, but retrieving the last submission and seeing if it is graded seems to be the right logic. 

I am a bit unsure about my decision, as Jonatan implemented the same logic that you are recommending.

Reply:
Just to confirm, I went a very hacky route here, so I would like to find an alternative to this too.

Reply:
Went a CTE route here, which works much better.